README.md
Persist-Json is a new fast Json parser written in Scala.
It includes:

   1. A standard Scala tree form of Json based on Scala immutable collections.
   2. A very fast Json parser that takes a Json string and produces the Scala form.
   3. A very fast Compact unparser that takes the Scala form and produces a compact single-line Json string.
   4. A Pretty unparser that takes the Scala form and produces a nicely formatted Json string.
   5. A set of methods for working with the Scala form that augment the existing standard collection methods.
   6. A mapper that converts between the Scala form and user-defined classes.

## Documentation
1. [ScalaDoc](http://nestorpersist.github.com/json)

## Reference

Persist-Json can be referenced in sbt from Maven Central as

    "com.persist" % "persist-json_2.12" % "1.2.0"
    
    "com.persist" % "persist-json_2.11" % "1.1.3"

## History
This parser started as part of the OStore NoSQL database
project.

[OStore](https://github.com/nestorpersist/ostore)

OStore makes extensive use of Json and needed a pure-Scala
Json parser whose output would be immutable Scala data.
For this purpose, the Twitter parser based on the parsing
combinator example in the Odersky book was choosen.

[Twitter Json](https://github.com/stevej/scala-json)

Although it met the functional need, it was much too slow.
The OStore Json parser mostly maintains the API of the 
Twitter Json parser but is very much faster.

Since the OStore Json parser has utility beyond its use in
OStore, it was broken off as this separate Persist-Json GitHub project.

## Thanks

YourKit is kindly supporting the Persist-Json open source project with its full-featured Java Profiler.
YourKit, LLC is the creator of innovative and intelligent tools for profiling
Java and .NET applications. Take a look at YourKit's leading software products:
[YourKit Java Profiler](http://www.yourkit.com/java/profiler/index.jsp) and
[YourKit .NET Profiler](http://www.yourkit.com/.net/profiler/index.jsp).




changes.txt
11/4/2014 Version 1.2.0
   UPgraded for Scala 2.12
10/8/2016 Version 1.1.4
   Fixed bug in Pretty with indent specified.
3/16/2016 Version 1.1.3
   Fixed ToJson mapper bugs for List and Vector.
3/12/2016 Version 1.1.2
   Updated dependency versions
1/8/2016 Version 1.1.1
   Fixed bug in compact sort
7/27/2015 Version 1.1.0
   Added sort parameter to Compact.
   Compact is now faster.
7/2/2015 Version 1.0.1
   Added Array read and write codecs.
6/30/2015 Version 1.0.0
   Upgraded all versions.
   Added some missing mapper codecs.
   Replaced the old reflective mapper with the new
   codec based mapper.
   Added some missing ScalaDoc.
3/9/2015
   Double Pretty and Compact now always use E notation.
8/26/2014
   Added safe option to Compact and Pretty.
   Adjusted decimal versus exp floating pt string reps.
8/25/2014
   Upgrade to Scala 2.11
   Add new to/from Json codex
3/24/2014 0.16
   Fixed bug in Pretty options.
3/22/2014 0.15
   Use Scala 2.10.3
   Use SBT 0.13.0
   Added more options to Pretty
7/27/2013 0.14
   Fix up scaladoc formating problems
   Add mapper support for short
   Better error reporting on ToObject
   Fixed bug in jput
5/13/2012 0.13
   Mapper now allows _ for key word escapes
   (for example, use constructor name _type for Json field "type")
   Mapper now supports case class parameters of type Option[Seq[T]]
5/8/2013 0.12
   Fixed bug in mapper toJson method for Option types.
5/2/2013 0.11
   Added jnull
   Use Scala 2.10.1
4/19/2012 0.10
   Compact and Pretty speed improvements
2/16/2013 0.9
   Compact and Pretty no longer escape "/"
   Compact and jget now accept mutable maps
1/19/2013 0.9-RC4
   performance improvements in parser
   new jgetBigDecimal and jgetDouble methods
12/18/2012 0.9-RC3
   fixed number parse bug introduced in RC2
   fixed multithread sync problem in object mapper
   added unit tests
12/15/2012 0.9-RC2
   Performance improvements


src/test/scala/com/persist/JsonTest.scala
/*
 *  Copyright 2012-2013 Persist Software
 *  
 *   http://www.persist.com
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
*/

package com.persist

import org.junit.runner.RunWith
import org.scalatest.junit.JUnitRunner
import JsonOps._
import org.scalatest.FunSuite

@RunWith(classOf[JUnitRunner])
class JsonTest extends FunSuite {

  test("json") {

    val s = """
      {
        "a":3,          // a comment
        "as":4,
        "b":"foo/bar",
        c:17,           // quotes not needed on simple field names
        "d":[1,2,3,4,5],
        "e":false,
        e1:true,
        e2:null,
        "f":-3,
        "g":2.3,
        "h":3e4, 
        "h1":-3E-4,
        i:""{
          A "raw string"
          with /multiple\ lines.
          }""
      }
            """
    val j = Json(s)
    jput(j, "js")(5.toShort) // test short value
    val c = Compact(j)
    val p = Pretty(j,indent=2)
    println(p)
    val j1 = Json(c)
    assert(jget(j,"h").getClass== jget(j1,"h").getClass) // make sure both are double
    assert(j == j1, "same tree")
    assert(c == Compact(j1), "same string")

    val A = jfield("a")
    val B = jfield("b")
    val C = jfield("c")
    val F1 = jfield(1)
    val F2 = jfield(2)

    val a: Json = JsonArray(2, 3, "a")
    val m: Json = JsonObject("a" -> 3, "b" -> 4)
    val n = JsonObject("a" -> m, "c" -> a)

    assert(jgetInt(m, "a") == 3, "getInt")
    assert(jgetLong(m, "a") == 3L, "getLong")

    assert(jsize(a) == 3, "array size")
    assert(jget(a, 0) == 2, "a(0)")
    assert(jget(a, 1) == 3, "a(1)")
    assert(jget(a, 2) == "a", "a(2)")

    assert(jsize(m) == 2, "object size")
    assert(jget(m, "a") == 3, "m.a")
    assert(jget(m, "b") == 4, "m.a")

    assert(jgetString(a, 2) == "a", "get array string")
    assert(jgetInt(m, "a") == 3, "get object int")
    assert(jget(n, "c", 0) == 2, "multilevel get")

    n match {
      case A(B(a: Int)) & C(F2(b: String)) => {
        assert(a == 4 && b == "a", "pattern problem")
      }
      case x => assert(false, "pattern fail")
    }

    assert(jput(emptyJsonObject, "a")(3) === JsonObject("a" -> 3), "put failed")
    assert(jput(emptyJsonObject, "a", "b")(3) === JsonObject("a" -> JsonObject("b" -> 3)), "put failed")
  }
}

src/test/scala/com/persist/MapperTest.scala
/*
 *  Copyright 2012-2015 Persist Software
 *  
 *   http://www.persist.com
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
*/

package com.persist

import org.junit.runner.RunWith
import org.scalatest.junit.JUnitRunner
import org.scalatest.FunSuite
import com.persist.JsonOps._
import com.persist.JsonMapper._
import com.persist.json._

case class Person(name: String, age: Option[Int], friend: Option[Person])

case class Group(city: String, people: Seq[Person], kinds:Array[String],
                 var cnt: Int, props: JsonObject, value: BigDecimal)

@RunWith(classOf[JUnitRunner])
class MapperTest extends FunSuite {

  test("mapper") {


    val j: Json = Json( """{city:"Seattle", cnt:2, props:{i:1, j:2}, value:2.3,
                         people:[{name:"Joe", friend:{name:"Sam"}},
                                 {name:"Tom", age:20}], kinds:["red","blue"]
                        }""")

    val group = ToObject[Group](j)

    val j1: Json = ToJson(group)

    assert(j1 === j, "mapper fail")
  }

}


src/test/scala/com/persist/PersonTest.scala
/*
 *  Copyright 2012-2015 Persist Software
 *
 *   http://www.persist.com
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
*/

package com.persist

import org.junit.runner.RunWith
import org.scalatest.junit.JUnitRunner
import org.scalatest.FunSuite
import com.persist.JsonOps._
//import com.persist.JsonMapper._
import java.nio.ByteBuffer
import com.persist.json.{ReadWriteCodec, ReadCodec, WriteCodec}

case class Person1(name: String, age: Option[Int])
case class Person(name: String, age: Option[Int], friend: Option[Person])
case class Group(city: String, people: Seq[Person], var cnt: Int, props: JsonObject, value:BigDecimal)


@RunWith(classOf[JUnitRunner])
class PersonTest extends FunSuite {
  implicit val rr = ReadCodec[Person1]
  implicit val pr = ReadCodec[Person]
  implicit val pw = WriteCodec[Person]
  implicit val gr = ReadCodec[Group]
  implicit val gw = WriteCodec[Group]

  test("mapper") {


    val j: Json = Json("""{city:"Seattle", cnt:2, props:{i:1, j:2}, value:2.3,
                         people:[{name:"Joe", friend:{name:"Sam"}},
                                 {name:"Tom", age:20}]
                        }""")

    //val mv = ToObject[Map[String,Json]](Map("a"->3,"b"->"foo"))
    //val iv = ToObject[Integer](17)
    val group: Group = json.read[Group](j)
    val j1: Json = json.toJson(group)

    assert(j1 === j, "mapper fail")
  }

}

index.js
Index.PACKAGES = {"com" : [], "com.persist" : [{"object" : "com\/persist\/Exceptions$.html", "name" : "com.persist.Exceptions"}, {"object" : "com\/persist\/JsonMapper$.html", "name" : "com.persist.JsonMapper"}, {"object" : "com\/persist\/JsonOps$.html", "name" : "com.persist.JsonOps"}], "com.persist.json" : [{"trait" : "com\/persist\/json\/package$$LowPriorityReadCodec.html", "name" : "com.persist.json.LowPriorityReadCodec"}, {"object" : "com\/persist\/json\/package$$ReadCodec$.html", "trait" : "com\/persist\/json\/package$$ReadCodec.html", "name" : "com.persist.json.ReadCodec"}, {"object" : "com\/persist\/json\/package$$ReadWriteCodec$.html", "trait" : "com\/persist\/json\/package$$ReadWriteCodec.html", "name" : "com.persist.json.ReadWriteCodec"}, {"object" : "com\/persist\/json\/package$$WriteCodec$.html", "trait" : "com\/persist\/json\/package$$WriteCodec.html", "name" : "com.persist.json.WriteCodec"}]};

src/main/scala/com/persist/Smile.scala
package com.persist

import java.io._
import com.persist.Exceptions.SystemException
import com.persist.JsonOps._
import scala.annotation.tailrec
import scala.collection.mutable.ArrayBuilder

object Smile {

  private[this] val PREFIX1 = 0x3A.toByte
  private[this] val PREFIX2 = 0x29.toByte
  private[this] val PREFIX3 = 0x0A.toByte
  private[this] val PREFIX4 = 0x00.toByte
  private[this] val EMPTY_STRING = 0x22.toByte
  private[this] val NULL = 0x21.toByte
  private[this] val FALSE = 0x22.toByte
  private[this] val TRUE = 0x23.toByte
  private[this] val START_OBJECT = 0xFA.toByte
  private[this] val END_OBJECT = 0xFB.toByte
  private[this] val START_ARRAY = 0xF8.toByte
  private[this] val END_ARRAY = 0xF9.toByte
  private[this] val STRING64 = 0x80.toByte
  private[this] val START_STRING = 0xE1.toByte
  private[this] val END_STRING = 0xFC.toByte
  private[this] val INT = 0x24.toByte
  private[this] val LONG = 0x25.toByte
  private[this] val END = 0xFF.toByte
  private[this] val UTF8 = java.nio.charset.Charset.forName("UTF-8")

  // TODO better exceptions
  case class SmileException() extends SystemException("Smile", "")

  def stringToBytes(s: String): Array[Byte] = {

    val ss = new ByteArrayOutputStream()
    val x = new OutputStreamWriter(ss, UTF8)
    x.write(s)
    x.close()
    ss.toByteArray
  }

  def bytesToString(b: Array[Byte]): String = {
    val ss = new ByteArrayInputStream(b)
    val x = new InputStreamReader(ss, UTF8)
    val sb = new StringBuilder(10)
    @tailrec def read(): Unit = {
      val i = x.read()
      if (i != -1) {
        sb.append(i.toChar)
        read()
      }
    }
    read()
    x.close()
    sb.toString()
  }

  def intToZigZag(n: Int): Int = (n << 1) ^ (n >> 31)

  def zigZagToInt(z: Int): Int = (z >> 1) ^ (-(z & 1))

  @tailrec def intToVInt7(n: Int, accum: Array[Byte]): Array[Byte] = {
    if (n == 0) {
      accum
    } else {
      val last7 = (n & 0x7F).toByte
      val rest = n >> 7
      intToVInt7(rest, last7 +: accum)
    }
  }

  def intToVInt(n: Int): Array[Byte] = {
    val last6 = (n & 0x3F).toByte
    val rest = n >> 6
    intToVInt7(rest, Array(last6))
  }

  // TODO
  def longToVInt(n: Long): Array[Byte] = {
    // at least 5 bytes
    ???
  }

  def longToZigZag(n: Long): Long = (n << 1) ^ (n >> 63)

  def zigZagToLong(z: Long): Long = (z >> 1) ^ (-(z & 1))

  def dump(a: Array[Byte]): Unit = {
    val s = for (i <- a) yield {
      val high = (i & 0xF0) >> 4
      val low = i & 0x0F
      f"$high%H$low%H"
    }
    println(s.mkString(" "))
  }

  def check(s: String) = {
    try {
      dump(ToBinary(Json(s)))
    } catch {
      case ex: Throwable =>
        println(ex)
    }
  }

  def ToBinary(j: Json): Array[Byte] = {
    val ab = ArrayBuilder.make[Byte]

    def add(b: Byte): Unit = ab += b

    def adda(b: Array[Byte]): Unit = ab ++= b

    def toBinary(j: Json) {
      j match {
        case s: String => {
          val b = stringToBytes(s)
          val size = b.size
          if (size == 0) {
            add(EMPTY_STRING)
          } else if (size < 64) {
            add((STRING64 + size).toByte)
            adda(b)
          } else {
            add(START_STRING)
            adda(b)
            add(END_STRING)
          }
        }
        case m: scala.collection.Map[_, _] => {
          add(START_OBJECT)
          for ((k, v) <- m) {
            toBinary(k)
            toBinary(v)
          }
          add(END_OBJECT)
        }
        case list: Seq[_] => {
          add(START_ARRAY)
          for (l <- list) toBinary(l)
          add(END_ARRAY)
        }
        case x: Int =>
          add(INT)
          val z = intToZigZag(x)
          val v = intToVInt(z)
          adda(v)
        case x: Long =>
          add(LONG)
          val z = longToZigZag(x)
          val v = longToVInt(z)
          adda(v)
        case x: BigDecimal => // TODO
        case x: Double => // TODO
        case x: Boolean => add(if (x) TRUE else FALSE)
        case x: Float => // TODO
        case x: Number => //TODO dispatch to int long, double or float, bigdecimal
        case null => add(NULL)
        case x => throw new SmileException()
      }
    }

    add(PREFIX1)
    add(PREFIX2)
    add(PREFIX3)
    add(PREFIX4)
    toBinary(j)
    add(END)
    ab.result()
  }

  def FromBinary(a: Array[Byte]): Json = {
    val it = a.toIterator
    var pos = 0;
    def getNext() = {
      if (it.hasNext) {
        pos += 1
        it.next
      } else {
        0xFF.toByte
      }
    }
    @tailrec def getObj(accum: JsonObject): Json = {
      val next = getNext()
      if (next == END_OBJECT) {
        accum
      } else {
        val k = fromBinary(next) match {
          case s: String => s
          case _ => throw SmileException()
        }
        val next1 = getNext()
        val v = fromBinary(next1)
        val accum1 = accum + (k -> v)
        getObj(accum1)
      }
    }

    def getString(): String = {
      // TODO
      ???
    }

    def getStringCnt(size: Int): String = {
      val b = for (i <- (0 until size).toArray) yield getNext()
      val s = bytesToString(b)
      s
    }

    def fromBinary(next: Byte): Json = {
      next match {
        case EMPTY_STRING => ""
        case NULL => jnull
        case FALSE => false
        case TRUE => true
        case START_OBJECT =>
          getObj(emptyJsonObject)
        case START_ARRAY => ???
        case START_STRING =>
          getNext()
          getString()
        case x =>
          val prefix = next & 0xF0
          prefix match {
            case 0X80 | 0x90 | 0xA0 | 0XB0 =>
              getStringCnt(next - STRING64)
            case x1 =>
              dump(Array(x))
              println(pos)
              throw SmileException()
          }
      }
    }
    if (getNext() != PREFIX1) throw SmileException()
    if (getNext() != PREFIX2) throw SmileException()
    if (getNext() != PREFIX3) throw SmileException()
    if (getNext() != PREFIX4) throw SmileException()
    val result = fromBinary(getNext())
    if (getNext() != END) throw SmileException()
    result
  }
}


src/main/scala/com/persist/JsonParse.scala
/*
 *  Copyright 2012-2013 Persist Software
 *  
 *   http://www.persist.com
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
*/

/*
 * This class is based on the Json parser given in the Odersky Scala book
 * as modified by Twitter.
 *     https://github.com/stevej/scala-json/
 * 
 * That version was however too slow, so it has been rewritten here with
 * a similar API, but with an emphasis on performance.
 * 
 */

package com.persist

import java.io.StringWriter

import scala.collection.immutable.{ListMap, HashMap}
import JsonOps._
import scala.util.Sorting
import Exceptions._
import scala.annotation.switch

private[persist] object JsonParse {

  // *** Character Kinds

  final type CharKind = Int
  final val Letter = 0
  final val Digit = 1
  final val Minus = 2
  final val Quote = 3
  final val Colon = 4
  final val Comma = 5
  final val Lbra = 6
  final val Rbra = 7
  final val Larr = 8
  final val Rarr = 9
  final val Blank = 10
  final val Other = 11
  final val Eof = 12
  final val Slash = 13

  // *** Token Kinds

  final type TokenKind = Int
  final val ID = 0
  final val STRING = 1
  final val NUMBER = 2
  final val BIGNUMBER = 3
  final val FLOATNUMBER = 4
  final val COLON = 5
  final val COMMA = 6
  final val LOBJ = 7
  final val ROBJ = 8
  final val LARR = 9
  final val RARR = 10
  final val BLANK = 11
  final val EOF = 12

  // *** Character => CharKind Map ***

  final val charKind1 = new Array[CharKind](256)

  for (i <- 0 until 255) {
    charKind1(i) = Other
  }
  for (i <- 'a'.toInt to 'z'.toInt) {
    charKind1(i) = Letter
  }
  for (i <- 'A'.toInt to 'Z'.toInt) {
    charKind1(i) = Letter
  }
  for (i <- '0'.toInt to '9'.toInt) {
    charKind1(i) = Digit
  }
  charKind1('-'.toInt) = Minus
  charKind1(','.toInt) = Comma
  charKind1('"'.toInt) = Quote
  charKind1(':'.toInt) = Colon
  charKind1('{'.toInt) = Lbra
  charKind1('}'.toInt) = Rbra
  charKind1('['.toInt) = Larr
  charKind1(']'.toInt) = Rarr
  charKind1(' '.toInt) = Blank
  charKind1('\t'.toInt) = Blank
  charKind1('\n'.toInt) = Blank
  charKind1('\r'.toInt) = Blank
  charKind1('/'.toInt) = Slash

  // *** Character Escapes

  final val escapeMap1 = HashMap[Int, String](
    '\\'.toInt -> "\\",
    '/'.toInt -> "/",
    '\"'.toInt -> "\"",
    'b'.toInt -> "\b",
    'f'.toInt -> "\f",
    'n'.toInt -> "\n",
    'r'.toInt -> "\r",
    't'.toInt -> "\t")

  def parse(s: String): Json = {
    val jp = new JsonParse(s)
    val result = jp.parse()
    result
  }
}

private[persist] class JsonParse(s: String) {

  // *** Import Shared Data ***

  import JsonParse._

  final private[this] val charKind = charKind1
  final private[this] val escapeMap = escapeMap1

  // *** INPUT STRING ***

  // array faster than accessing string directly using charAt
  //final private[this] val s1 = s.toCharArray()
  final private[this] val size = s.size

  // *** CHARACTERS ***

  final private[this] var pos = 0

  final private[this] var ch: Int = 0
  final private[this] var chKind: CharKind = 0
  final private[this] var chLinePos: Int = 0
  final private[this] var chCharPos: Int = 0

  final private def chNext {
    if (pos < size) {
      //ch = s1(pos).toInt
      ch = s.charAt(pos)
      chKind = if (ch < 255) {
        charKind(ch)
      } else {
        Other
      }
      pos += 1
      if (ch == '\n'.toInt) {
        chLinePos += 1
        chCharPos = 1
      } else {
        chCharPos += 1
      }
    } else {
      ch = -1
      pos = size + 1
      chKind = Eof
    }
  }


  final private def chError(msg: String): Nothing = {
    throw new JsonParseException(msg, s, chLinePos, chCharPos)
  }

  final private def chMark = pos - 1

  final private def chSubstr(first: Int, delta: Int = 0) = {
    s.substring(first, pos - 1 - delta)
  }

  // *** LEXER ***

  final private[this] var tokenKind: TokenKind = BLANK
  final private[this] var tokenValue: String = ""
  final private[this] var linePos = 1
  final private[this] var charPos = 1

  final private def getDigits() {
    while (chKind == Digit) {
      chNext
    }
  }

  final private def handleDigit() {
    val first = chMark
    getDigits()
    val k1 = if (ch == '.'.toInt) {
      chNext
      getDigits()
      BIGNUMBER
    } else {
      NUMBER
    }
    val k2 = if (ch == 'E'.toInt || ch == 'e'.toInt) {
      chNext
      if (ch == '+'.toInt) {
        chNext
      } else if (ch == '-'.toInt) {
        chNext
      }
      getDigits()
      FLOATNUMBER
    } else {
      k1
    }
    tokenKind = k2
    tokenValue = chSubstr(first)
  }

  final private def handleRaw() {
    chNext
    var first = chMark
    var state = 0
    do {
      if (chKind == Eof) chError("EOF encountered in raw string")
      state = if (ch == '}') {
        1
      } else if (ch == '"') {
        if (state == 1) {
          2
        } else if (state == 2) {
          3
        } else {
          0
        }
      } else {
        0
      }
      chNext
    } while (state != 3)
    tokenKind = STRING
    tokenValue = chSubstr(first, 3)
  }

  final private def tokenNext {
    do {
      linePos = chLinePos
      charPos = chCharPos
      val kind: Int = chKind
      (kind: @switch) match {
        case Letter => {
          val first = chMark
          while (chKind == Letter || chKind == Digit) {
            chNext
          }
          tokenKind = ID
          tokenValue = chSubstr(first)
        }
        case Digit => handleDigit()
        case Minus => {
          chNext
          handleDigit()
          tokenValue = "-" + tokenValue
        }
        case Quote => {
          var sb: StringBuilder = null
          chNext
          var first = chMark
          while (ch != '"'.toInt && ch >= 32) {
            if (ch == '\\'.toInt) {
              if (sb == null) sb = new StringBuilder(50)
              sb.append(chSubstr(first))
              chNext
              escapeMap.get(ch) match {
                case Some(s) => {
                  sb.append(s)
                  chNext
                }
                case None => {
                  if (ch != 'u'.toInt) chError("Illegal escape")
                  chNext
                  var code = 0
                  for (i <- 1 to 4) {
                    val ch1 = ch.toChar.toString
                    val i = "0123456789abcdef".indexOf(ch1.toLowerCase)
                    if (i == -1) chError("Illegal hex character")
                    code = code * 16 + i
                    chNext
                  }
                  sb.append(code.toChar.toString)
                }
              }
              first = chMark
            } else {
              chNext
            }
          }
          if (ch != '"') chError("Unexpected string character:" + ch.toChar)
          val s1 = chSubstr(first)
          val s2 = if (sb == null) s1
          else {
            sb.append(s1)
            sb.toString
          }
          tokenKind = STRING
          tokenValue = s2
          chNext
          if (s2.length() == 0 && ch == '{') {
            handleRaw()
          }
        }
        case Colon => {
          chNext
          tokenKind = COLON
          tokenValue = ""
        }
        case Comma => {
          chNext
          tokenKind = COMMA
          tokenValue = ""
        }
        case Lbra => {
          chNext
          tokenKind = LOBJ
          tokenValue = ""
        }
        case Rbra => {
          chNext
          tokenKind = ROBJ
          tokenValue = ""
        }
        case Larr => {
          chNext
          tokenKind = LARR
          tokenValue = ""
        }
        case Rarr => {
          chNext
          tokenKind = RARR
          tokenValue = ""
        }
        case Blank => {
          do {
            chNext
          } while (chKind == Blank)
          tokenKind = BLANK
          tokenValue = ""
        }
        case Other => chError("Unexpected character")
        case Eof => {
          chNext
          tokenKind = EOF
          tokenValue = ""
        }
        case Slash => {
          val first = chMark
          if (chKind != Slash) chError("Expecting Slash")
          do {
            chNext
          } while (ch != '\n' && chKind != Eof)
          tokenKind = BLANK
          tokenValue = ""
        }
      }
    } while (tokenKind == BLANK)
  }

  final private def tokenError(msg: String): Nothing = {
    throw new JsonParseException(msg, s, linePos, charPos)
  }

  // *** PARSER ***

  final private def handleEof() {
    tokenError("Unexpected eof")
  }

  final private def handleUnexpected() {
    tokenError("Unexpected input")
  }

  final private def handleArray() = {
    tokenNext
    var result = List[Json]()
    while (tokenKind != RARR) {
      val t = getJson
      result = t +: result
      if (tokenKind == COMMA) {
        tokenNext
      } else if (tokenKind == RARR) {
      } else {
        tokenError("Expecting , or ]")
      }
    }
    tokenNext
    result.reverse
  }

  final private[this] val emptyMap = HashMap[String, Json]()

  final private def handleObject() = {
    tokenNext
    var result = emptyMap
    while (tokenKind != ROBJ) {
      if (tokenKind != STRING && tokenKind != ID) tokenError("Expecting string or name")
      val name = tokenValue
      tokenNext
      if (tokenKind != COLON) tokenError("Expecting :")
      tokenNext
      val t = getJson
      result += (name -> t)
      if (tokenKind == COMMA) {
        tokenNext
      } else if (tokenKind == ROBJ) {
      } else {
        tokenError("Expecting , or }")
      }
    }
    tokenNext
    result
  }

  final private def getJson(): Json = {
    val kind: Int = tokenKind
    val result = (kind: @switch) match {
      case ID => {
        val result = if (tokenValue == "true") {
          true
        } else if (tokenValue == "false") {
          false
        } else if (tokenValue == "null") {
          null
        } else {
          tokenError("Not true, false, or null")
        }
        tokenNext
        result
      }
      case STRING => {
        val result = tokenValue
        tokenNext
        result
      }
      case NUMBER => {
        val v = try {
          tokenValue.toLong
        } catch {
          case _: Throwable => tokenError("Bad integer")
        }
        tokenNext
        val r: Json = if (v >= Int.MinValue && v <= Int.MaxValue) v.toInt else v
        r
      }
      case BIGNUMBER => {
        val v = try {
          BigDecimal(tokenValue)
        } catch {
          case _: Throwable => tokenError("Bad decimal number")
        }
        tokenNext
        v
      }
      case FLOATNUMBER => {
        val v = try {
          tokenValue.toDouble
        } catch {
          case _: Throwable => tokenError("Bad double")
        }
        tokenNext
        v
      }
      case COLON => handleUnexpected()
      case COMMA => handleUnexpected()
      case LOBJ => handleObject()
      case ROBJ => handleUnexpected()
      case LARR => handleArray()
      case RARR => handleUnexpected()
      case EOF => handleEof()
    }
    result
  }

  final def parse(): Json = {
    chNext
    tokenNext
    val result = getJson
    if (tokenKind != EOF) tokenError("Excess input")
    result
  }
}

private[persist] object JsonUnparse {

  def quote(sb: StringBuilder, s: String) {
    s.foreach {
      case ch =>
        (ch: @switch) match {
          case '\r' => sb.append("\\r")
          case '\n' => sb.append("\\n")
          case '\t' => sb.append("\\t")
          case '"' => sb.append("\\\"")
          case '\\' => sb.append("\\\\")
          case '\u00ff' => sb.append("\u00ff")
          case c =>
            if (ch < ' ') {
              if (c <= 0xF) sb.append("\\u000").append(Integer.toHexString(c))
              else if (c <= 0xFF) sb.append("\\u00").append(Integer.toHexString(c))
              else if (c <= 0xFFF) sb.append("\\u0").append(Integer.toHexString(c))
              else sb.append("\\u").append(Integer.toHexString(c))
            } else {
              sb.append(ch)
            }
        }
    }
  }


  def compact(obj: Json, safe: Boolean, sort: Boolean): String = {
    val sb = new StringBuilder(500)
    def compact1(obj1: Json) {
      obj1 match {
        case s: String => {
          sb.append("\"")
          quote(sb, s)
          sb.append("\"")
        }
        case m: scala.collection.Map[_, _] => {
          if (m.isEmpty) {
            sb.append("{}")
          } else {
            val m3 = if (sort) {
              val m2 = m.asInstanceOf[scala.collection.Map[String, Json]].iterator.toList
              ListMap(m2.toSeq.sortBy(_._1): _*)
            } else {
              m
            }
            var sep = "{"
            for ((name, elem) <- m3) {
              sb.append(sep)
              sb.append("\"")
              quote(sb, name.asInstanceOf[String])
              sb.append("\":")
              compact1(elem)
              sep = ","
            }

            sb.append("}")
          }
        }
        case list: Seq[_] => {
          if (list.isEmpty) {
            sb.append("[]")
          } else {
            var sep = "["
            for (elem <- list) {
              sb.append(sep)
              compact1(elem)
              sep = ","
            }
            sb.append("]")
          }
        }
        case x: Int => sb.append(x.toString)
        case x: Long => sb.append(x.toString)
        case x: BigDecimal => sb.append(x.toString)
        case x: Double =>
          sb.append("%1$e".format(x)) // g->e
        case x: Boolean => sb.append(x.toString)
        case x: Float =>
          sb.append("%1$e".format(x)) // g->e
        case x: Number => sb.append(x.toString)
        case null => sb.append("null")
        case x => if (safe) {
          val bad = JsonObject("BAD1" -> x.toString)
          compact1(bad)
        } else {
          throw new SystemException("JsonUnparse", JsonObject("msg" -> "bad json value", "value" -> x.toString()))
        }
      }
    }
    compact1(obj)
    sb.toString
  }

  private def isMultiLine(s: String): Boolean = s.indexOf("\n") >= 0

  private def doIndent(s: String, indent: Int, first: String = ""): String = {
    val space = " " * indent
    if (isMultiLine(s)) {
      val parts = s.split("\n")
      val head = parts.head
      val tail = parts.tail
      val indent1 = first.size + indent
      val space1 = " " * indent1
      val head1 = space + first + head
      val tail1 = tail.map(part => space1 + part)
      val seq1 = head1 +: tail1
      seq1.mkString("\n")
    } else {
      space + first + s
    }
  }

  private def wrap(first: String, sep: String, last: String, indent: Int, seq: Seq[String]): String = {
    if (seq.isEmpty) {
      doIndent(first + last, indent)
    } else {
      val indent1 = first.size + indent
      val head = seq.head
      val tail = seq.tail
      val head1 = doIndent(head, indent, first)
      val tail1 = tail.map(part => doIndent(part, indent1))
      val seq1 = head1 +: tail1
      seq1.mkString(",\n") + "\n" + doIndent(last, indent)
    }
  }

  private def split(s: Seq[String], width: Int, count: Int): Boolean = {
    s.size > count ||
      s.map(isMultiLine(_)).fold(false) {
        _ || _
      } ||
      s.map(_.size).fold(0)(_ + _) + s.size + 2 > width
  }

  /**
   * Returns a pretty JSON representation of the given object
   */
  def pretty(obj: Json, indent: Int, width: Int, count: Int, safe: Boolean): String = {
    obj match {
      case null => doIndent("null", indent)
      case x: Boolean => doIndent(x.toString, indent)
      case x: Double => doIndent("%1$e".format(x), indent) // g=>e
      case x: Float => doIndent("%1$e".format(x), indent) // g=>e
      case x: Number => doIndent(x.toString, indent)
      case array: Array[Json] => pretty(array.toList, indent, width, count, safe)
      case list: Seq[_] =>
        val strings = list.map(pretty(_, 0, width, count, safe))
        if (!split(strings, width, count)) {
          doIndent("[" + strings.mkString(",") + "]", indent)
        } else {
          wrap("[", ",", "]", indent, strings)
        }
      case map: scala.collection.Map[_, _] =>
        val seq2 = Sorting.stableSort[(Any, Json), String](map.iterator.toList, {
          case (k, v) => k.toString
        })
        val strings = seq2.map {
          case (k, v) => {
            val v1 = pretty(v, 0, width, count, safe)
            val sb = new StringBuilder(500)
            sb.append("\"")
            quote(sb, k.toString)
            sb.append("\":")
            val label = sb.toString
            if (isMultiLine(v1) || label.size + v1.size > width) {
              label + "\n" + doIndent(v1, 2)
            } else {
              label + v1
            }
          }
        }
        if (!split(strings, width, count)) {
          doIndent("{" + strings.mkString(",") + "}", indent)
        } else {
          wrap("{", ",", "}", indent, strings)
        }
      case s: String =>
        val sb = new StringBuilder(500)
        sb.append("\"")
        quote(sb, s)
        sb.append("\"")
        doIndent(sb.toString, indent)
      case x =>
        if (safe) {
          val bad = JsonObject("BAD" -> x.toString)
          pretty(bad, indent, width, count, safe)
        } else {
          throw new SystemException("JsonUnparse", JsonObject("msg" -> "bad json value", "value" -> x.toString()))
        }
    }
  }
}

